File: erykandroidcommon/src/main/java/pl/edu/amu/wmi/erykandroidcommon/recycler/MyRecyclerViewAdapter.java
package pl.edu.amu.wmi.erykandroidcommon.recycler;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

class MyRecyclerViewAdapter<T extends UniqueItem, S extends AbstractViewHolder<T>> extends RecyclerView.Adapter<S> {

    private final AbstractFragmentGrid<T, S> abstractFragmentGrid;

    private List<T> values = new ArrayList<>();

    MyRecyclerViewAdapter(AbstractFragmentGrid<T, S> abstractFragmentGrid) {
        this.abstractFragmentGrid = abstractFragmentGrid;
    }

    void setValues(List<T> values) {
        this.values = values;
    }

    void addValue(T value) {
        values.add(value);
    }


    @Override
    public S onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(abstractFragmentGrid.getItemViewId(), parent, false);
        return abstractFragmentGrid.createViewHolder(view);
    }

    @Override
    public void onBindViewHolder(S holder, int position) {
        abstractFragmentGrid.setListenerRow(holder, values.get(position));
    }

    @Override
    public int getItemCount() {
        return values.size();
    }

    public T getItemByPosition(int position) {
        return values.get(position);
    }

    public void updateValue(T item) {
        for (int i = 0; i < values.size(); ++i) {
            T listItem = values.get(i);
            if (listItem.getId() == item.getId()) {
                values.set(i, item);
            }
        }
    }

    public void delete(int id) {
        for (int i = 0; i < values.size(); ++i) {
            T listItem = values.get(i);
            if (listItem.getId() == id) {
                values.remove(i);
            }
        }
    }
}

File: src/main/java/pl/edu/amu/wmi/erykandroidcommon/recycler/MyRecyclerViewAdapter.kt
package pl.edu.amu.wmi.erykandroidcommon.recycler

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import java.util.ArrayList

public class MyRecyclerViewAdapter<T : UniqueItem, S : AbstractViewHolder<T>>(private val abstractFragmentGrid: AbstractFragmentGrid<T, S>) : RecyclerView.Adapter<S>() {

    private var values: MutableList<T> = ArrayList()

    fun setValues(values: MutableList<T>) {
        this.values = values
    }

    fun addValue(value: T) {
        values.add(value)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): S {
        val view = LayoutInflater.from(parent.context).inflate(abstractFragmentGrid.itemViewId, parent, false)
        return abstractFragmentGrid.createViewHolder(view)
    }

    override fun onBindViewHolder(holder: S, position: Int) {
        abstractFragmentGrid.setListenerRow(holder, values[position])
    }

    override fun getItemCount(): Int = values.size

    fun getItemByPosition(position: Int): T = values[position]

    fun updateValue(item: T) {
        for (i in values.indices) {
            val listItem = values[i]
            if (listItem.id == item.id) {
                values[i] = item
            }
        }
    }

    fun delete(id: Int) {
        for (i in values.indices) {
            val listItem = values[i]
            if (listItem.id == id) {
                values.removeAt(i)
            }
        }
    }
}
